pkg/controller/sshd/sshd_controller.go

```diff
@@ -0,0 +1,564 @@
+
+	// Fetch the SSHD instance
+	instance := &cloudingressv1alpha1.SSHD{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	// Ensure we have an awsClient instance.
+	if r.awsClient == nil {
+		region, err := utils.GetClusterRegion(r.client)
+		if err != nil {
+			r.SetSSHDStatusError(instance, reqLogger, "Failed to get cluster's AWS region", err)
+			return reconcile.Result{}, err
+		}
+		r.awsClient, err = awsclient.GetAWSClient(r.client, awsclient.NewAwsClientInput{
+			SecretName: config.AWSSecretName,
+			NameSpace:  config.OperatorNamespace,
+			AwsRegion:  region,
+		})
+		if err != nil {
+			r.SetSSHDStatusError(instance, reqLogger, "Failed to create an AWS client", err)
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Check for a deletion timestamp.
+	if instance.DeletionTimestamp.IsZero() {
+		// Request object is alive, so ensure it has the DNS finalizer.
+		if !controllerutil.ContainsFinalizer(instance, reconcileSSHDFinalizerDNS) {
+			controllerutil.AddFinalizer(instance, reconcileSSHDFinalizerDNS)
+			if err = r.client.Update(context.TODO(), instance); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	} else {
+		// Request object is being deleted.
+		if controllerutil.ContainsFinalizer(instance, reconcileSSHDFinalizerDNS) && r.route53 != nil {
+			r.SetSSHDStatus(instance, "Deleting DNS aliases", cloudingressv1alpha1.SSHDStateFinalizing)
+			if err = r.deleteDNSRecords(); err != nil {
+				return reconcile.Result{}, err
+			}
+
+			// Remove the DNS finalizer and update the request object.
+			controllerutil.RemoveFinalizer(instance, reconcileSSHDFinalizerDNS)
+			if err = r.client.Update(context.TODO(), instance); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+
+		// Halt the reconciliation.
+		return reconcile.Result{}, nil
+	}
+
+	// List ConfigMaps with SSH keys
+	configMapList := &corev1.ConfigMapList{}
+	selector, err := metav1.LabelSelectorAsSelector(&instance.Spec.ConfigMapSelector)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if err = r.client.List(context.TODO(), configMapList,
+		client.InNamespace(instance.Namespace),
+		&client.MatchingLabelsSelector{Selector: selector}); err != nil {
+		r.SetSSHDStatusError(instance, reqLogger, "Failed to list config maps with SSH keys", err)
+		return reconcile.Result{}, err
+	}
+	// Sort ConfigMaps by name for the purpose of creating
+	// stable Volume and VolumeMount lists in the Deployment.
+	sort.Slice(configMapList.Items, func(i, j int) bool {
+		return configMapList.Items[i].Name < configMapList.Items[j].Name
+	})
```

Comment:
This bit seems to be wanting a custom resource to represent a SSH key. I'd love to see that so we can talk about SSH keys and not config maps.

Reply:
Separately, I'm curious why this approach was taken instead of a single ConfigMap with many keys inside of it used to represent each associate's key, eg:

```yaml
data:
  alice: ssh-rsa ...
  bob: ssh-rsa ...
  barack: ssh-rsa ...
```

It seems like this would be a cleaner effect and the be less work to build up a new list each time during reconciliation and has the added benefit that when mounted inside the container, it would be [kept up to date by Kubernetes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically). (Though I still like the idea of a separate custom resource for each associate; this single ConfigMap has merit as well.)

What do you think Matt?

Reply:
-1 to a CRD for each associate. We don't need to spend a watch every individual associate, we don't need to maintain fields beyond the key, and the only use for this at this point is to drop it into the FS of the container. A config map or secret meets all of these requirements without a new custom data type.

Is there any particular benefit to splitting up into individual files for every associate, as opposed to one authorized_keys file per configmap? The only reason we have to split these into separate configmaps in the first place, is that not all associates get access to all clusters (e.g. CSSRE)

Reply:
The intention between @luis-falcon and I was to generate one configmap and selectorsyncset per team.  Jenkins jobs will generate these objects on Hive.  The configmap will contain all the SSH keys for that team and the selectorsyncset will target clusters appropriate for that team.  (e.g. everything for SREP, but only RHMI clusters for CS-SRE, etc.)

The controller will find whatever teams' configmaps are present on the cluster and prepare a volume/volumemount for each in the pod spec of the deployment.